### PR TITLE
Fix Default missing some default values

### PR DESCRIPTION
**Pull request by alexarchambault**
Fixes https://github.com/milessabin/shapeless/issues/474

### CODE REVIEW

**Comment by ceedubs** [comment]
@alexarchambault should these tests be using `assertTypedEquals` instead of ascribing the type (which could influence inference)?

I assume that the return types can be inferred.

**Comment by alexarchambault** [comment]
Good point, it'd be better to ensure we get the right type without influencing inference.

I'm not too keen about using `assertTypedEquals[A]` for that though, as it only checks that the type under scrutiny is `<: A`, not that it is `=:= A`. I'd like to be sure that the `Some` parameters aren't literal ones, e.g. I want `Some[String]`, not ``Some[Witness.`"b"`.T]`` (note that this point is not tested currently, just saying it would be better like this).

There's a [`equalInferredType` function](https://github.com/milessabin/shapeless/blob/master/core/src/test/scala/shapeless/hlist.scala#L450) that does just that. So here, I think I'd prefer to devise a `assertSameTypeEquals`, ensuring that the inferred types of both of its arguments are the same (and not that they are just `<:` to a common super type).

**Comment by ceedubs** [comment]
I just recalled that I think we are reliving a past conversation :) https://github.com/milessabin/shapeless/pull/97#issuecomment-45450057

This PR seems to fix the much bigger problem of default arguments not working for top-level classes, so I'd hate for it to get hung up on being finicky about testing the exact inferred types. I say :+1: for either merging it now or with `assertTypedEquals` and/or `equalInferredType`, and we can always revisit our typed equal test functions.

**Comment by alexarchambault** [comment]
@ceedubs I remember https://github.com/milessabin/shapeless/pull/97 :-) I just used `assertTypedEquals` in the tests of `Default`, so that it would easier to revisit them later.

**Comment by ceedubs** [comment]
:+1:

**Comment by alexarchambault** [comment]
Who else could :+1: that... :smile: @travisbrown ?

File: core/src/test/scala/shapeless/default.scala [hunk 1]
```diff
@@ -4,6 +4,12 @@ import shapeless.record.Record
 
 import org.junit.Test
 import shapeless.test.illTyped
+import shapeless.testutil.assertTypedEquals
+
+// Willingly defined in the shapeless scope - (compile time) reflection API not behaving the same way
```
Comment (travisbrown) [inline core/src/test/scala/shapeless/default.scala]: This is nitpicky, but the important thing is that this is a top-level class, not that it's in the `shapeless` package.

Even more nitpicky: "intentionally" would be more idiomatic than "willingly".

**Comment by travisbrown** [comment]
:+1:, looks good to me!

**Comment by alexarchambault** [comment]
Changed the above comment following @travisbrown advice.

As the last change is rather minor, I'll consider this good to merge in a few hours unless anyone objects.

**Comment by alexarchambault** [comment]
(Forced push a change in a test method name (now `topLevel`), that should have been done along with the previous change.)

**Comment by milessabin** [comment]
Thanks folks :-)

**Pull request merged**